File: 2. Inorder traversal recursive.cpp
/*
 Task: Inorder traversal (recursive)
 Time Complexity: O(n)
 Space Complexity: O(h) — recursion stack, where h is tree height
*/
#include <bits/stdc++.h>

using namespace std;

struct Node
{
    int data;
    Node* left;
    Node* right;

    Node(int d) : data(d), left(nullptr), right(nullptr) {}
};

void inorder(Node* root)
{
    if (!root)
    {
        return;
    }

    inorder(root->left);
    cout << root->data << " ";
    inorder(root->right);
}

int main()
{
    Node* root = new Node(1);
    root->left = new Node(2);
    root->right = new Node(3);
    root->left->left = new Node(4);
    root->left->right = new Node(5);

    cout << "Inorder: ";
    inorder(root);
    cout << endl;
    return 0;
}

// sc tc

File: 2. Inorder traversal recursive.py
"""
Task: Inorder traversal (recursive)
Time Complexity: O(n)
Space Complexity: O(h) — recursion stack
"""

class Node:
    def __init__(self,data): self.data=data; self.left=None; self.right=None

def inorder(root):
    if not root: return
    inorder(root.left)
    print(root.data, end=' ')
    inorder(root.right)

if __name__=='__main__':
    root=Node(1); root.left=Node(2); root.right=Node(3); root.left.left=Node(4); root.left.right=Node(5)
    print('Inorder:', end=' ')
    inorder(root); print()
